### PR TITLE
Call.cancel at any time.

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

**Comment by swankjesse** [comment]
(this is consistent with OkHttp)

**Comment by baoti** [comment]
I like this.

**Comment by felipecsl** [comment]
good stuff :+1:

File: retrofit/src/main/java/retrofit/OkHttpCall.java [hunk 1]
```diff
@@ -144,11 +155,11 @@ private void callSuccess(Response<T> response) {
     }
   }
 
-  public void cancel() {
+  public synchronized void cancel() {
```
Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/OkHttpCall.java]: I hate that the use of this ends up in the docs.

File: retrofit/src/main/java/retrofit/OkHttpCall.java [hunk 2]
```diff
@@ -30,6 +30,7 @@
 
   private volatile com.squareup.okhttp.Call rawCall;
   private boolean executed; // Guarded by this.
+  private boolean canceled; // Guarded by this.
```
Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/OkHttpCall.java]: Can this just be a volatile? We never read and write in one place. One person just writes and another just reads.

File: retrofit/src/main/java/retrofit/OkHttpCall.java [hunk 3]
```diff
@@ -30,6 +30,7 @@
 
   private volatile com.squareup.okhttp.Call rawCall;
   private boolean executed; // Guarded by this.
```
Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/OkHttpCall.java]: Which makes me think this could use an AtomicBooleanFieldUpdater for lock-free CAS changes

**Comment by JakeWharton** [comment]
LGTM. Nice tests.

File: retrofit/src/main/java/retrofit/OkHttpCall.java (hunk 1 above)
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/OkHttpCall.java]: You mean `synchronized`? It doesn't

File: retrofit/src/main/java/retrofit/OkHttpCall.java (hunk 2 above)
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/OkHttpCall.java]: yup, though I prefer `synchronized` 'cause it's easier to reason about

**Pull request merged**